File: python/142.py
#142. Given a linked list, return the node where the cycle begins. If there is no cycle, return null.
# There is a cycle in a linked list if there is some node in the list that can be reached again 
# by continuously following the next pointer. Internally, pos is used to denote the index of the node 
# that tail's next pointer is connected to. Note that pos is not passed as a parameter.
# Notice that you should not modify the linked list.

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def detectCycle(head: ListNode) -> ListNode:
        if(head == None or head.next == None):
            return None
        slow = head
        fast = head
        entry = head
        while(fast.next and fast.next.next):
            slow = slow.next
            fast = fast.next.next
            if(slow == fast):
                while(slow != entry):
                    slow = slow.next
                    entry = entry.next
                return entry
        return None

if __name__ == "__main__":
    #ex. if cycle begin at pos 1
    head = ListNode(3)
    a = ListNode(2)
    b = ListNode(0)
    c = ListNode(-4)
    c.next = a
    a.next = b
    b.next = c
    head.next = a

    print(detectCycle(head))

File: golang/142.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func (n ListNode) PrintList() {
	for ; n.Next != nil; n = *n.Next {
		fmt.Print(n.Val, "->")
	}
	fmt.Print(n.Val)
	fmt.Println()
}

func main() {
	c := ListNode{Val: -4}
	b := ListNode{0, &c}
	a := ListNode{2, &b}
	c.Next = &a
	head := ListNode{3, &a}
	//head.PrintList()
	detectCycle(&head).PrintList()
}

// DESC: using two pointers, faster and slower by using this two pointers technique
// we can determined that if theres a cycle on the linkedlist
// these two pointers will be on the same node at some point.
// Next we need to define where is the cycle occurs. by running
// the slow pointer and the entry pointer. these two will meet at where the cycle occurs
func detectCycle(head *ListNode) *ListNode {
	if &head == nil && &head.Next == nil {
		return nil
	}
	fast, slow, entry := head, head, head
	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
		if fast == slow {
			for slow != entry {
				slow = slow.Next
				entry = entry.Next
			}
			return entry
		}
	}
	return nil
}
